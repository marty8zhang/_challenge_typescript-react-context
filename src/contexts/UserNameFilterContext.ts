import React from 'react';
import User from '../models/User';

export interface UserNameFilterContextValue {
  userName: string,
  handleUserNameChange: (event: React.ChangeEvent<HTMLInputElement>) => void,
  users: User[],
  filteredUsers: User[],
}

const UserNameFilterContext = React.createContext<UserNameFilterContextValue>({
  userName: '',
  handleUserNameChange: () => {},
  users: [],
  filteredUsers: [],
});

export { UserNameFilterContext as default };
