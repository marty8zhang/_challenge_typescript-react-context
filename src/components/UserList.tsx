import React, { useContext } from 'react';
import UserNameFilterContext from '../contexts/UserNameFilterContext';

export default function UserList() {
  const { filteredUsers } = useContext(UserNameFilterContext);
  const userList = filteredUsers.map((user) => (
    <tr key={user.id}>
      <td>{user.name}</td>
      <td>{user.email}</td>
      <td>{user.address.city}</td>
      <td>{user.company.name}</td>
    </tr>
  ));

  return (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Email</th>
          <th>City</th>
          <th>Company</th>
        </tr>
      </thead>
      <tbody>
        {userList.length ? userList
          : (
            <tr>
              <td colSpan={4}>No User Found.</td>
            </tr>
          )}
      </tbody>
    </table>
  );
}
