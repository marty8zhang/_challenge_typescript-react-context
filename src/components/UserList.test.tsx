import React from 'react';
import { render, screen } from '@testing-library/react';
import UserList from './UserList';
import UserNameFilterContext from '../contexts/UserNameFilterContext';

it('displays a table of the retrieved users', async () => {
  const users = [
    {
      id: 1,
      name: 'Name 1',
      email: 'email1@example.com',
      address: {
        city: 'City 1',
      },
      company: {
        name: 'Company 1',
      },
    },
    {
      id: 2,
      name: 'Name 2',
      email: 'email2@example.com',
      address: {
        city: 'City 2',
      },
      company: {
        name: 'Company 2',
      },
    },
    {
      id: 3,
      name: 'Name 3',
      email: 'email3@example.com',
      address: {
        city: 'City 3',
      },
      company: {
        name: 'Company 3',
      },
    },
  ];

  const { container } = render(
    <UserNameFilterContext.Provider value={{
      userName: '',
      handleUserNameChange: jest.fn(),
      users,
      filteredUsers: [users[0], users[2]],
    }}
    >
      <UserList />
    </UserNameFilterContext.Provider>,
  );

  const table = container.firstChild;
  expect(table!.textContent).toContain('Name 1');
  expect(table!.textContent).not.toContain('Name 2');
  expect(table).toMatchSnapshot();
});

it('displays a message in the table when no user is retrieved', async () => {
  const { container } = render(
    <UserNameFilterContext.Provider value={{
      userName: '',
      handleUserNameChange: jest.fn(),
      users: [],
      filteredUsers: [],
    }}
    >
      <UserList />
    </UserNameFilterContext.Provider>,
  );
  const message = screen.getByText('No User Found.');
  expect(message).toBeInTheDocument();

  const table = container.firstChild;
  expect(table).toMatchSnapshot();
});
