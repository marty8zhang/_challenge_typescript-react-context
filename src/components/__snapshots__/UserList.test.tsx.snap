// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a message in the table when no user is retrieved 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th>
        City
      </th>
      <th>
        Company
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colspan="4"
      >
        No User Found.
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`displays a table of the retrieved users 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th>
        City
      </th>
      <th>
        Company
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Name 1
      </td>
      <td>
        email1@example.com
      </td>
      <td>
        City 1
      </td>
      <td>
        Company 1
      </td>
    </tr>
    <tr>
      <td>
        Name 3
      </td>
      <td>
        email3@example.com
      </td>
      <td>
        City 3
      </td>
      <td>
        Company 3
      </td>
    </tr>
  </tbody>
</table>
`;
